### ISSUE TITLE
Regent: Allow variables in projection functors

### ISSUE THREAD

**Issue by manopapad** [StanfordLegion/legion#474]
Currently Regent will reject this program:

```
import 'regent'

task foo(r : region(ispace(int1d),float))
where reads(r) do
end

task main()
  var r = region(ispace(int1d,10),float)
  var N = 2
  var colors = ispace(int1d,N)
  var p_r = partition(equal, r, colors)
  var k = 1
  __demand(__parallel)
  for c in colors do
    foo(p_r[(c + k) % N])
  end
end
```

with the error message:

```
test.rg:15: loop optimization failed: argument 1 is not provably projectable or invariant
    foo(p_r[(c + k) % N])
       ^
```

This seems to be happening because `k` and `N` are variables declared outside the index-space loop, and the codegen for projection functors is not handling those.

**Comment by elliottslaughter** [comment]
The reason we can't do this right now is that there isn't a good way to pass parameters to a projection functor. In theory, we need to either:

1. Dynamically register a new projection functor for every index launch, or
2. Register once and figure out another way to pass in the arguments

I don't particularly like (1), registration is not something we should be doing inside the body of the main loop (and if we do decide this is the way to go, we need to make sure this registration is very efficient or else it will really hurt).

One way to implement (2) would be to pass any free variables in via e.g. task arugments. This may be problematic under dynamic control replication where, as I understand it, the runtime memoizes the result of the projection functor if `is_functional` set to true. Strictly speaking, these projection functors would still be pure, but if Legion isn't aware of the extra arguments to the projection functor (e.g. if it doesn't memoize based on task arguments) then from Legion's perspective it will violate the purity assumption.

@lightsighter Can you advise on one option would be best? How much will it cost us if we set `is_functional` to false and get Legion DCR to disable its memoization?

**Comment by lightsighter** [comment]
The runtime isn't memoizing projection functor results in the control replication branch anymore. Doing away with composite views allowed us to get rid of the code that needs to invert the projection functor which was expensive and the reason we needed to memoize in the first place.

**Comment by manopapad** [comment]
Recording a discussion I had with @magnatelee about this yesterday. Feel free to correct me if I misunderstood, I'm not very familiar with how index-space launches are handled under the covers.

An issue with alternative (2) that wasn't obvious to me is that we don't have control over the timing of the task launches. So assume that we have a loop like this:

```
for k = 0,10 do
  for c in colors do
    foo(p_r[c + k])
  end
end
```

and we naively code-gen it based on alternative (2):

```
var pf = lambda x. x + a
for k = 0,10 do
  pf.a = k
  index_launch(task=foo, ispace=colors, partition=p_r, proj=pf)
end
```

With this code, the control loop can run ahead and schedule multiple iterations of the outer loop before the first index-space launch happens, which will cause the first few launches of `foo` to use the wrong value of `a` on the projection functor.

Therefore, to implement (2) properly we would need to insert a fence before `pf.a = k`, which means we inserted a spurious block in the middle of the control loop, which is possibly worse overhead than (1).

**Comment by elliottslaughter** [comment]
If you're suggesting we would literally use a closure for the projection functor---that's not what I had in mind. That wouldn't even work on multiple nodes (because the closure values wouldn't get sync'd between nodes), and projection functors need to be valid everywhere.

My suggestion was going to be to stuff the extra parameters into the global task arg of the index launch. Right now Regent only uses local task args---we literally run the loop and pack local task args for each point in the launch domain. So global task args are currently unused.

I was going to propose a new calling convention where the free variables in a projection functor are identified. When you code generate an index launch, you take all the free variables from all the projection functors in the launch, and insert them all into a big struct. Then you code generate the projection functor to decode the struct and fish out the values it wants.

In this way, the arguments for a projection functor are saved at launch time and if they get modified later, it's no big deal, just like normal task args.

**Comment by elliottslaughter** [comment]
I was thinking about this again recently, and one issue with the approach I suggested is that it doesn't generalize well to index operations other than task launches (because IndexCopy and IndexFill don't have task arguments). I see a couple options:

 1. Don't support variables in projection functors for non-IndexTasks. (This makes support inconsistent across the language, but I can't actually come up with any cases we'd want such a projection functor in an IndexFill or IndexCopy.)
 2. Add a TaskArgument to IndexCopy / IndexFill, which is ignored by the runtime and therefore would only be used by projection functors and/or possibly mappers.
 3. Change the projection functor API so that each time a projection functor is used it can be accompanied by an optional TaskArgument, which is then passed when the projection functor is evaluated (but is not otherwise used by the runtime).
 4. Switch to dynamic projection functor registration. I don't like this solution because we'll potentially be registering a projection functor on each iteration of the main simulation loop (and potentially multiple times per iteration). Registration becomes a blocking operation (in DCR, we have to use a barrier across all shards to make sure no one gets ahead). Things break outright if we're not doing DCR across the whole machine (e.g. in any non-DCR code, or in DCR code where the set of shards is smaller than the whole machine). In addition, the data lifetimes are very non-obvious (due to asynchronous execution); the closure has to be heap allocated and it's not even obvious whether it can be deleted by the end of the task (if the task completes before the children operations---possibly can be fixed by an execution fence but that's just more ugliness).

I'd prefer (2) or (3) but I think (1) is probably the most expedient answer.

**Comment by lightsighter** [comment]
Why do you need to pass a TaskArgument to your projection functors? Most of the non-deprecated versions of the projection functor API don't take them as arguments.
https://gitlab.com/StanfordLegion/legion/blob/master/runtime/legion.h#L3131-3188

**Comment by elliottslaughter** [comment]
I'd like to be able to index-launch the inner loop in codes like:

```
for i = 0, N do
  __demand(__parallel)
  for j = i + 1, N do
    my_task(some_part[{i, j}])
  end
end
```

This example is sort of trivial but the same sort of pattern shows up in `language/examples/cholesky.rg`.

Point is, the index expression `{i, j}` captures a variable `j` which is invariant for the duration of the loop but not for the duration of the program. I don't really want to generate a literal closure, for the reasons outlined in (4) above, and I don't want to register `N` copies of the projection functor. I'd really prefer to find some way to pass an argument to the projection functor, and at the moment (for task launches only, not copies and fills) the most expedient way to do that appears to be (1) above and to use the task's global task argument for projection functor data.

**Comment by elliottslaughter** [comment]
We discussed this at the meeting today, and the proposal we came up with was to:

  * Add a `TaskArgument` field to the `RegionRequirement` that represents a blob of bits associated with the `RegionRequirement` that get copied around along with it and are available to projection functors, mappers, etc.
  * Do not otherwise change the projection functor interface. (You can get to the `TaskArgument` by looking up the region requirement by its index in the `Task` or other operation object.)

This does slightly increase the size of the `RegionRequirement` struct, but this is a struct that's already relatively large so the relative impact should not be that large. Actual data will be copied only when it's present, so the cost of the `TaskArgument` itself should be minimal.

The change should not be an API breaking change because we're not modifying any function signatures, just adding a field to the struct.

Thoughts welcome.

**Comment by lightsighter** [comment]
I've added support now for "projection args" on region requirements. You can add them and get them off the region requirements.

https://gitlab.com/StanfordLegion/legion/commit/63241c02051f061bfee3fe2b497ecf4ff04e8e05

Let me know if you have any problems with them.

I've left up to you guys to figure out how to put this in the C API.

**Comment by elliottslaughter** [comment]
This is fixed in master as of commit 3c002953c0e3d19cd5a0d61e1820e016832d4578.
